### PR TITLE
Switch from sjcl to noble-curves and noble-hashes

### PR DESCRIPTION
**Pull request by paulmillr**
I have removed `InnerScalar` hack because `private` property access (s.k) may fail at some point in the future when typescript will start compiling it to actual private property.

Benchmark results on M2 / node v18.14.0. Before:

```
OPRF/P256_SHA256/blind     x 43.97 ops/sec ±0.10% (72 runs sampled)
OPRF/P256_SHA256/blindEval x 121 ops/sec ±0.23% (83 runs sampled)
OPRF/P256_SHA256/finalize  x 121 ops/sec ±0.17% (83 runs sampled)
OPRF/P384_SHA384/blind     x 31.72 ops/sec ±0.25% (77 runs sampled)
OPRF/P384_SHA384/blindEval x 93.83 ops/sec ±0.12% (88 runs sampled)
OPRF/P384_SHA384/finalize  x 92.29 ops/sec ±0.18% (87 runs sampled)
OPRF/P521_SHA512/blind     x 14.72 ops/sec ±0.11% (72 runs sampled)
OPRF/P521_SHA512/blindEval x 46.10 ops/sec ±0.16% (75 runs sampled)
OPRF/P521_SHA512/finalize  x 44.52 ops/sec ±4.16% (74 runs sampled)
VOPRF/P256_SHA256/blind     x 44.08 ops/sec ±0.36% (72 runs sampled)
VOPRF/P256_SHA256/blindEval x 15.83 ops/sec ±2.53% (76 runs sampled)
VOPRF/P256_SHA256/finalize  x 17.95 ops/sec ±2.55% (17 runs sampled)
VOPRF/P384_SHA384/blind     x 30.59 ops/sec ±2.33% (74 runs sampled)
VOPRF/P384_SHA384/blindEval x 11.87 ops/sec ±3.11% (59 runs sampled)
VOPRF/P384_SHA384/finalize  x 14.56 ops/sec ±0.66% (16 runs sampled)
VOPRF/P521_SHA512/blind     x 14.37 ops/sec ±1.42% (70 runs sampled)
VOPRF/P521_SHA512/blindEval x 6.02 ops/sec ±0.57% (34 runs sampled)
VOPRF/P521_SHA512/finalize  x 12.10 ops/sec ±12.39% (15 runs sampled)
POPRF/P256_SHA256/blind     x 44.12 ops/sec ±0.10% (72 runs sampled)
POPRF/P256_SHA256/blindEval x 14.00 ops/sec ±0.54% (69 runs sampled)
POPRF/P256_SHA256/finalize  x 15.67 ops/sec ±0.54% (24 runs sampled)
POPRF/P384_SHA384/blind     x 31.66 ops/sec ±0.29% (76 runs sampled)
POPRF/P384_SHA384/blindEval x 10.39 ops/sec ±0.49% (53 runs sampled)
POPRF/P384_SHA384/finalize  x 11.62 ops/sec ±0.19% (19 runs sampled)
POPRF/P521_SHA512/blind     x 14.69 ops/sec ±0.11% (72 runs sampled)
POPRF/P521_SHA512/blindEval x 4.95 ops/sec ±0.72% (28 runs sampled)
POPRF/P521_SHA512/finalize  x 6.82 ops/sec ±0.33% (15 runs sampled)
P-256/add          x 1,321 ops/sec ±0.26% (99 runs sampled)
P-256/mulgen       x 123 ops/sec ±0.13% (80 runs sampled)
P-256/mul          x 122 ops/sec ±0.25% (80 runs sampled)
P-256/hashToScalar x 15,058 ops/sec ±0.22% (89 runs sampled)
P-256/hashToGroup  x 258 ops/sec ±0.68% (87 runs sampled)
P-384/add          x 923 ops/sec ±0.25% (99 runs sampled)
P-384/mulgen       x 92.53 ops/sec ±0.61% (81 runs sampled)
P-384/mul          x 93.20 ops/sec ±0.22% (81 runs sampled)
P-384/hashToScalar x 9,452 ops/sec ±0.55% (89 runs sampled)
P-384/hashToGroup  x 183 ops/sec ±0.26% (87 runs sampled)
P-521/add          x 409 ops/sec ±0.40% (96 runs sampled)
P-521/mulgen       x 44.35 ops/sec ±2.40% (59 runs sampled)
P-521/mul          x 45.12 ops/sec ±2.44% (61 runs sampled)
P-521/hashToScalar x 5,979 ops/sec ±0.90% (87 runs sampled)
P-521/hashToGroup  x 81.83 ops/sec ±0.27% (79 runs sampled)
```

After:

```
OPRF/P256_SHA256/blind     x 376 ops/sec ±0.18% (89 runs sampled)
OPRF/P256_SHA256/blindEval x 424 ops/sec ±0.26% (90 runs sampled)
OPRF/P256_SHA256/finalize  x 416 ops/sec ±0.22% (88 runs sampled)
OPRF/P384_SHA384/blind     x 158 ops/sec ±0.21% (84 runs sampled)
OPRF/P384_SHA384/blindEval x 179 ops/sec ±0.33% (84 runs sampled)
OPRF/P384_SHA384/finalize  x 177 ops/sec ±0.22% (92 runs sampled)
OPRF/P521_SHA512/blind     x 82.92 ops/sec ±0.33% (80 runs sampled)
OPRF/P521_SHA512/blindEval x 94.02 ops/sec ±0.39% (89 runs sampled)
OPRF/P521_SHA512/finalize  x 92.93 ops/sec ±0.39% (88 runs sampled)
VOPRF/P256_SHA256/blind     x 375 ops/sec ±0.20% (92 runs sampled)
VOPRF/P256_SHA256/blindEval x 100 ops/sec ±0.23% (80 runs sampled)
VOPRF/P256_SHA256/finalize  x 77.26 ops/sec ±0.27% (91 runs sampled)
VOPRF/P384_SHA384/blind     x 158 ops/sec ±0.26% (92 runs sampled)
VOPRF/P384_SHA384/blindEval x 42.37 ops/sec ±0.39% (69 runs sampled)
VOPRF/P384_SHA384/finalize  x 32.92 ops/sec ±0.30% (79 runs sampled)
VOPRF/P521_SHA512/blind     x 82.71 ops/sec ±0.31% (79 runs sampled)
VOPRF/P521_SHA512/blindEval x 22.33 ops/sec ±0.54% (57 runs sampled)
VOPRF/P521_SHA512/finalize  x 17.49 ops/sec ±0.47% (83 runs sampled)
POPRF/P256_SHA256/blind     x 376 ops/sec ±0.16% (92 runs sampled)
POPRF/P256_SHA256/blindEval x 100 ops/sec ±0.13% (80 runs sampled)
POPRF/P256_SHA256/finalize  x 75.90 ops/sec ±0.22% (89 runs sampled)
POPRF/P384_SHA384/blind     x 158 ops/sec ±0.18% (92 runs sampled)
POPRF/P384_SHA384/blindEval x 42.39 ops/sec ±0.29% (69 runs sampled)
POPRF/P384_SHA384/finalize  x 32.05 ops/sec ±0.39% (77 runs sampled)
POPRF/P521_SHA512/blind     x 82.72 ops/sec ±0.32% (79 runs sampled)
POPRF/P521_SHA512/blindEval x 22.21 ops/sec ±0.56% (56 runs sampled)
POPRF/P521_SHA512/finalize  x 17.03 ops/sec ±0.37% (81 runs sampled)
P-256/add          x 257,802 ops/sec ±0.16% (99 runs sampled)
P-256/mulgen       x 5,365 ops/sec ±0.07% (102 runs sampled)
P-256/mul          x 423 ops/sec ±0.25% (95 runs sampled)
P-256/hashToScalar x 83,171 ops/sec ±0.08% (92 runs sampled)
P-256/hashToGroup  x 3,320 ops/sec ±0.18% (92 runs sampled)
P-384/add          x 173,196 ops/sec ±0.61% (99 runs sampled)
P-384/mulgen       x 2,430 ops/sec ±0.66% (98 runs sampled)
P-384/mul          x 175 ops/sec ±0.55% (91 runs sampled)
P-384/hashToScalar x 44,492 ops/sec ±2.01% (86 runs sampled)
P-384/hashToGroup  x 1,364 ops/sec ±0.19% (93 runs sampled)
P-521/add          x 126,775 ops/sec ±0.30% (100 runs sampled)
P-521/mulgen       x 1,324 ops/sec ±0.18% (100 runs sampled)
P-521/mul          x 91.47 ops/sec ±0.46% (80 runs sampled)
P-521/hashToScalar x 44,770 ops/sec ±0.05% (93 runs sampled)
P-521/hashToGroup  x 686 ops/sec ±0.47% (90 runs sampled)
```

### CODE REVIEW

File: src/group.ts [hunk 1]
```diff
@@ -356,110 +182,19 @@ export class Elt {
...
-        return new Elt(g, point)
-    }
-
     static async hash(g: Group, msg: Uint8Array, dst: Uint8Array): Promise<Elt> {
-        const u = await Elt.hashToField(g, msg, dst, 2)
-        const Q0 = Elt.sswu(g, u[0])
-        const Q1 = Elt.sswu(g, u[1])
-        return Q0.add(Q1)
+        const h2c = getCurve(g.id).h2c
+        const DST = new TextDecoder().decode(dst)
```
Comment (armfazh) [inline src/group.ts:195]: `DST` must be an array of bytes, not a string. (In the h2c draft we used ascii strings for simplicity).

The change must come from here: https://github.com/paulmillr/noble-curves/blob/main/src/abstract/hash-to-curve.ts#L130

File: src/group.ts [hunk 2]
```diff
@@ -356,110 +182,19 @@ export class Elt {
...
-        y = y.mul(tv1)
-
-        const point = new sjcl.ecc.pointJac(curve, x, y, z).toAffine()
-        if (!point.isValid()) {
-            throw new Error('point not in curve')
-        }
-        return new Elt(g, point)
-    }
-
     static async hash(g: Group, msg: Uint8Array, dst: Uint8Array): Promise<Elt> {
```
Comment (armfazh) [inline src/group.ts:193]: This is not async.

File: src/group.ts [hunk 3]
```diff
@@ -539,3 +274,19 @@ export class Group {
         return Scalar.hash(this, msg, dst)
     }
 }
+
+type CurveDef = {
+    fp: Field<bigint>
+    Point: ProjConstructor<bigint>
+    h2c: typeof hashToP256
+}
+const curves: Record<GroupID, CurveDef> = {
```
Comment (armfazh) [inline src/group.ts:283]: are ristretto and decaf groups also supported in noble package? voprf specification also admits these groups.

File: src/group.ts [hunk 4]
```diff
@@ -152,200 +85,93 @@ export class Scalar {
...
+        const array = bytes.subarray(0, g.size)
+        const k = bytesToNumberBE(array)
+        const fp = getCurve(g.id).fp
+        if (k >= fp.ORDER) {
             throw errDeserialization(Scalar)
         }
         return new Scalar(g, k)
     }
 
     static async hash(g: Group, msg: Uint8Array, dst: Uint8Array): Promise<Scalar> {
```
[the remark this answers is not in the document]
Comment (paulmillr) [reply src/group.ts:97]: It returns promise so it is "async". I've kept old apis for compatibility.

File: src/group.ts (hunk 3 above)
Comment (paulmillr) [reply src/group.ts:283]: ristretto yeah

File: src/group.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulmillr) [reply src/group.ts:97]: Should I change this to sync or keep async for now?

[the remark this answers is not in the document]
Comment (armfazh) [reply src/group.ts:97]: it's ok to change it to sync. I thought the linter will catch this.

File: package.json [hunk 5]
```diff
@@ -53,6 +53,7 @@
         "format": "prettier './(src|test|bench|examples)/*.ts' --write"
     },
     "dependencies": {
-        "sjcl": "1.0.8"
+        "@noble/curves": "^0.7.2",
+        "@noble/hashes": "^1.2.0"
```
Comment (armfazh) [inline package.json:57]:
```suggestion
        "@noble/curves": "0.7.2",
        "@noble/hashes": "1.2.0"
```

**Review by cjpatton — COMMENTED** [review]
Thanks @paulmillr for the contribution! These look like great libraries. After having a closer look at noble-curves (I have not had a chance to look at noble-hashes yet), the selling points of these implementations are indeed quite attractive.

My initial concern is that there was no effort to make the field arithmetic constant-time. As the [README rightly points out](https://github.com/paulmillr/noble-curves#security), we can't reasonably expect a JS implementation of things like scalar multiplication to mitigate all possible side channels (e.g., an attacker that shares CPU cache with the process). That said, it *may* be reasonable to expect such code to resist less sophisticated threats (e.g., an attacker making requests to the process from the network). Put simply: orders of magnitude matter.

Before taking this PR, I'd like to suggest we assess whether there are regressions in side-channel resistance. In particular, if the SJCL made any effort to make the field arithmetic constant time, then I think we should keep this dependency for now.

I should emphasize however that I'm not really an expert on how to write low-level crypto securely (@armfazh is) and I'm not an expert in JS. Which leads me to a question: **What accounts for the speed up of noble-curves over SJCL?** (E.g., nearly 3.5x for P-256 scalar multiplication, if I'm reading your benchmarks correctly.) Perhaps there side-channel mitigations built into SJCL's field arithmetic that make it inherently slower?

**Comment by paulmillr** [comment]
No. Sjcl does not have ct mitigations. Noble curves do. Sjcl is basically unsupported buggy garbage these days.

**Comment by paulmillr** [comment]
> My initial concern is that there was no effort to make the field arithmetic constant-time

- [`inverseMod`](https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/bn.js#L186) is not CT
- [`power`](https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/bn.js#L270) is not CT
- [toAffine](https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/ecc.js#L239) is not CT
- [Jacobian Point add](https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/ecc.js#L173) is not CT! (noble is)
- [Jacobian Point double](https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/ecc.js#L218) is not CT! (noble is)
- [Jacobian Point multiplication](https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/ecc.js#L253) is not CT!!! (noble is) — **terrible**
- [ec sign](https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/ecc.js#L682) uses non-deterministic ECDSA, seems trivial to break with their in-house csprng (noble uses native csprng) — **terrible**

Bugs:

- [secp224k1 produces points which are not on-curve](https://github.com/bitwiseshiftleft/sjcl/issues/262)
- [bignum](https://github.com/bitwiseshiftleft/sjcl/issues/411) produces incorrect result for subtraction
- makes stuff like P256, P384 also incorrect

You're saying "noble is bad" because we don't have ct fields, however, noble has things which are a lot more important: algorithmically CT multiplication using second accumulator, complete addition formulas. Just do some EC ct tests, you'll see what is where.

**Comment by bwesterb** [comment]
I haven't studied the topic constant-time javascript in much detail yet, but it scares me. For instance, JITs commonly specialise functions on commonly used arguments, and that's a bummer if the argument is meant to be secret.

With that big caveat out of the way: noble relies on `BigInt`, which [does not have any constant time guarantees](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#cryptography). There are javascript libaries that claim constant-timeness. What if we swap out BigNum for the one from [this library](https://github.com/soatok/constant-time-js).

**Comment by paulmillr** [comment]
@bwesterb I wrote this paragraph on MDN.

**Comment by paulmillr** [comment]
Folks, you literally have dumpster fire right now with regards to timing attacks (curves themselves) and you're talking about some theoretical nonsense. Did you check the library you've linked? There is no CT inversion there. This doesn't make any sense.

**Comment by chris-wood** [comment]
Thanks @paulmillr for taking the time to prepare and send this PR! We're investigating this change and will get back to this thread when we have time. Please be patient. In the meantime, if you could clean up conflicts, that would help make our assessment easier. Please also be mindful of our [code of conduct](https://github.com/cloudflare/.github/blob/master/CODE_OF_CONDUCT.md) when engaging here. Let's keep the discussion and tone professional and focus on the technical matters at hand. We look forward to your continued, constructive contributions!

**Comment by paulmillr** [comment]
Understood.

To make decision easier for you, check out how sjcl calculation time depends on private key bits, effectively leaking all the timings:

```
sjcl private key A x 7,624 ops/sec @ 131μs/op ± 7.36% (min: 74μs, max: 313μs)
sjcl private key B x 117 ops/sec @ 8ms/op
sjcl private key C x 56 ops/sec @ 17ms/op
```

Reproducible with this code:

```js
// npm init -y && npm install micro-bmark sjcl-including-ecc
const bmark = require('micro-bmark');
const sjcl = require('sjcl-including-ecc');
const curve = sjcl.ecc.curves.k256;
// Generate new key
// var k2 = sjcl.ecc.ecdsa.generateKeys(curve, 0); console.log(k2.sec.serialize())
const privA = '1000000000000000000000000000000000000000000000000000000000000000';
const privB = '0000000000000000000000000000010000000000000000000000000000000000';
const privC = '0000000000000000000000000000000000000000000000000000000000000001';
bmark.run(async () => {
  console.log(curve.G.mult(privA).isIdentity);
  await bmark.mark('sjcl private key A', 110, () => curve.G.mult(privA));
  await bmark.mark('sjcl private key B', 110, () => curve.G.mult(privB));
  await bmark.mark('sjcl private key C', 110, () => curve.G.mult(privC));
})
```

This is catastrophically bad result and it won't happen with curves. By the way, curves got audited, I will publish it in the repo asap.

**Comment by paulmillr** [comment]
audit is live now:

https://github.com/trailofbits/publications/blob/master/reviews/2023-01-ryanshea-noblecurveslibrary-securityreview.pdf

**Comment by paulmillr** [comment]
It has been around 1 month since the PR was created. I have shown you how your current cryptography is not secure and is bleeding with vulnerability leaking timings that could be exploited today, given enough resources. On a contrary, curves were audited by a third party.

Is this going to be reviewed, or did I do some unnecessary investment of 7 hours of free work in cloudflare repository that you've agreed upon before?

This is not how open source is developed. Either you don't want it and say in advance, or at least leave some feedback to OSS contributors.

**Comment by chris-wood** [comment]
As I said above, @paulmillr, we are still investigating this PR amongst many other competing priorities. Please be patient. Being disrespectful will not expedite the process.

**Comment by paulmillr** [comment]
sjcl will be crushed by noble-curves any day now. Watch for weekly downloads and dependents who've made their pick at this url:

- https://www.npmjs.com/package/sjcl
- https://www.npmjs.com/package/@noble/curves

**Comment by sublimator** [comment]
This seems a clear improvement. An MIT licensed (can always fork) open source lib by a passionate author vs an ancient lib that requires vendoring and patching. Not to mention the clear performance gains!

**Comment by armfazh** [comment]
Update: in #35 , there is a new way to make the group implementation plug-able. This enable users to provide a different (than sjcl) implementation.

**Comment by sublimator** [comment]
@armfazh 

Would you be fine with bundling a @noble implementation in this package? (and updating the test files to test it) 

Can prefix it with experimental (to start?) 
Something like: "import { NobleGroup } from 'voprf-ts/experimental-noble-group'" 

Also, while I'm at it messing with package.json "exports", I would like to sort out commonjs support (see #30 )

**Pull request closed without merging**